src/odbc/src/app/application_data_buffer.cpp

```diff
@@ -260,10 +260,8 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
   if (ANSI_STRING_ONLY) {
     bytesRequired = value.length() * outCharSize;
   } else {
-    thread_local std::wstring_convert<std::codecvt_utf8<wchar_t >, wchar_t>
-      converter;
-    std::wstring inString = converter.from_bytes(value.c_str());
-    bytesRequired = inString.length() * outCharSize;
+    size_t maxExpansionFactor = 4; // Safe upper bound (utf-8 to utf-32)
+    bytesRequired = (value.length() * maxExpansionFactor + 1) * outCharSize;
```

Comment:
```suggestion
    bytesRequired = (std::mbstowcs(nullptr, value.c_str(), 0) * outCharSize);
```

Rather than using the maximum upper bound, we can use `mbstowcs` without a target buffer to avoid the initial conversion, and avoid the wide string copy all together.